core/src/main/java/io/questdb/cairo/O3PurgeDiscoveryJob.java

```diff
@@ -28,181 +28,266 @@
+        try {
+            txnScoreboard.ofRO(path);
+            path.trimTo(tableRootLen);
+            txReader.ofRO(path, partitionBy);
+            loadLastTx(txReader);
+
+            for (int i = 0; i < n; i += 2) {
+                if (partitionList.get(i + 1) != partitionTs) {
+                    if (i > lo + 2) {
+                        processPartition(
+                                ff,
+                                path,
+                                tableRootLen,
+                                txReader,
+                                txnScoreboard,
+                                partitionTs,
+                                partitionBy,
+                                partitionList,
+                                lo,
+                                i
+                        );
                     }
+                    lo = i;
+                    partitionTs = partitionList.get(i + 1);
                 }
             }
-            return true;
+            // Tail
+            if (n > lo + 2) {
+                processPartition(
+                        ff,
+                        path,
+                        tableRootLen,
+                        txReader,
+                        txnScoreboard,
+                        partitionTs,
+                        partitionBy,
+                        partitionList,
+                        lo,
+                        n
+                );
+            }
+        } catch (CairoException ex) {
+            // It is possible that table is dropped while this async job was in the queue.
+            // so it can be not too bad. Log error and continue work on the queue
+            LOG.error().$(ex.getFlyweightMessage()).$();
+        } finally {
+            txReader.clear();
+            txnScoreboard.clear();
         }
-        return false;
     }
 
-    private static void processDir(
-            MutableCharSink sink,
-            StringSink fileNameSink,
-            CharSequence tableName,
-            LongList txnList,
-            long pUtf8NameZ,
-            int type
-    ) {
-        if (Files.isDir(pUtf8NameZ, type, fileNameSink)) {
-            if (Chars.startsWith(fileNameSink, sink)) {
-                // extract txn from name
-                int index = Chars.lastIndexOf(fileNameSink, '.');
-                if (index < 0) {
-                    txnList.add(-1);
-                } else {
-                    try {
-                        txnList.add(Numbers.parseLong(fileNameSink, index + 1, fileNameSink.length()));
-                    } catch (NumericException e) {
-                        LOG.error().$("unknown directory [table=").utf8(tableName).$(", dir=").utf8(fileNameSink).$(']').$();
-                    }
-                }
+    private void parsePartitionDateVersion(StringSink fileNameSink, DirectLongList partitionList, CharSequence tableName, DateFormat partitionByFormat) {
+        int index = Chars.lastIndexOf(fileNameSink, '.');
```

Comment:
Multiple calls to .length() method, is it changing?

Reply:
fixed in https://github.com/questdb/questdb/pull/1778/commits/c07fd7aa136747a7679d3865ffbb164535d0ee3d